### ISSUE TITLE
file multiupload support

### ISSUE THREAD

**Issue by raneomik** [zenstruck/browser#52]
Greetings !

first of all, I really enjoy your library with its Developper eXperience friendly usage.

But we came to cases with multiple file upload in a Symfony application.
Is there a way to test a multiple upload on a file input with `multiple="multiple"` attribute ?   

this kind of "generic code" (`$files` being an array of paths to fixtures files) :

```php
    $browser = $this
        ->browser()
        ->get('/form-with-file-input-path')
     ;
     ...
     foreach($files as $file) {
         $brower->attachFile('file_input[]', $file);
     }

    $browser->click('submit_button');
```
doesn't seem to work : in the controller, the concerned file field, or the `$request->files` seem to be systematicaly empty ....

thank you !

EDIT: sorry,  in the controller, the concerned file field actually contains the last file path from the `$files` array. But it's problematic on forms with constraints like for example "at least 3 files are required"

**Comment by kbond** [comment]
Interesting timing, I was just thinking about this and if it would work. I guess it doesn't... Browser uses Mink under the hood and I found [this issue](https://github.com/minkphp/Mink/issues/358). It's possible mink doesn't support this but the issue isn't clear if that's the case...

Question: do you know how to do this using _vanilla_ Symfony DomCrawler/BrowserKit? I found [this section](https://symfony.com/doc/current/components/dom_crawler.html#forms) in the docs about file uploads but it doesn't show a multiple upload example.

**Comment by raneomik** [comment]
yes, we kind of resolved our multipload issue using the vanilla symfony this way : 

```php
        $client = static::createClient();
        $crawler = $client->request('GET', '/form-with-file-input-path');
        $form = $crawler->selectButton('submit')->form();

        $node = $crawler->filter("input[name='file_input[files][]']")->getNode(0);

        foreach ($files as $file) {
            $newField = (new FileFormField($node));
            $newField->upload($file);
            $form->set($newField);
        }

        $client->submit($form, $irrelevantHereOtherFormFieldValues);
```

it was adapted from this : https://stackoverflow.com/a/48068458

it's just those are the only cases in our functionnal test base that doesn't use the zenstruck/browser :sweat_smile:

**Comment by kbond** [comment]
Haha, I just found that SO article!

Ok, I might be able to add a hack to allow this.

**Comment by kbond** [comment]
To be clear, would you be cool if this was the API:

```php
$browser->attachFile('Upload', 'path/to/file.txt'); // single file (as now)

$browser->attachFile('Upload', ['path/to/file1.txt', 'path/to/file2.txt']); // array of files
```

**Comment by raneomik** [comment]
of course, maybe another method `attachFiles` / `attachMultipleFiles` to keep the typing distinction between `string` and `array` respectively in current `attachFile` and (not current) `attachFiles`/ `attachMultipleFiles` at second parameter ...? IMHO

**Comment by kbond** [comment]
When you have a chance, could you test #53 in your app for me?

BTW, I opted to keep the single `->attachFile()` method (for now). I feel too many similarly named methods can hurt discoverability when using your IDE's autocompletion to find the method you want. I might re-evaluate this before 1.0.

**Comment by raneomik** [comment]
The multipload works like a charm, thank you !

(its out of topic here, but we just had another problem when we tested it in our case - we have several buttons to validate the form ("publish" & "save as draft"), but when we check which one is clicked like this :

```php

    if ($form->isSubmitted() && $form->isValid())) {
        dd($form->getClickedButton());
   }
```

the output is `null`, without being warned that "Link with id|title|alt|text "save_as_draft" not found."

Should I open another issue about this ? Not sure about this but it seems to concern only the code I downloaded from your branch `kbond:attach-multiple-files` ...
)

**Comment by kbond** [comment]
Strange but I don't have a test case covering this currently. I'll add one and see if I can see the same issue.

For clarity, you have two submit input's with different `name` attributes? 

_Edit:_ It's been a while since I've added multiple submits and am unsure on how this is normally done. Looks like you're using Symfony Forms to add the buttons too?

I found some code in an app:

```html
<input type="submit" name="submit" value="Save" />
<button name="submit" value="save-show">Save and Show</button>
```

Then in my controller, I have:

```php
$request->request->get('submit'); // returns "Save" or "save-show" depending on which is clicked
```

Is this inline with your code?

**Comment by raneomik** [comment]
yes that's it, we add buttons in our FormType like this : 

```php
$builder
            ->add(self::SAVE_AS_DRAFT_BUTTON, SubmitType::class)
            ->add(self::PUBLISH_BUTTON, SubmitType::class)
;
```

and check the button in our controller : 

```php
        if ($form->isSubmitted() && $form->isValid() 
           && null !== ($data = $form->getData())
        ) {
            $submitType = $form->get(ExampleFormType::SAVE_AS_DRAFT_BUTTON);
            $data->setAsDraft($submitType->isClicked());
      
            ...
```

and we have functionnal tests to verify if `$data` is draft or not, like this : 

```php
        $this->browser()
            ->get('/form-path')
... // filling fields
            ->click('Save as draft') //translated button label
    ;

    $object = ObjectFactory::last();
    self::assertTrue($object->isDraft()); // fail

```

The name values of those buttons are indeed different - we'll put the submit button addition in the templates instead of in the FormType, that should resolve our issue.

Thanks a lot !

**Comment by kbond** [comment]
> we'll put the submit button addition in the templates instead of in the FormType, that should resolve our issue.

Let me know if this solves your issue but it is concerning that #53 broke your existing functionality. I did refactor some internal form handling logic so I could have broken something. I'll add some tests with different scenarios to confirm.

**Comment by kbond** [comment]
I think #54 demonstrates the different scenarios:
1. Two submit buttons with the same name and different values
2. Another submit button with a different name/value

~Once I rebase #53 to include I _should_ be able to find the issue.~

_Edit:_ damn, didn't find the issue you're experiencing...

**Comment by kbond** [comment]
@raneomik, I merged #53 and released [v0.7.0](https://github.com/zenstruck/browser/releases/tag/v0.7.0) with it included. Feel free to open another issue to discuss the button click issue further.

**Comment by raneomik** [comment]
okay, thank you !
